### ISSUE TITLE
Warn about "if FUNC; then" and "FUNC ||" that suppresses "-e" in the function

### ISSUE THREAD

**Issue by pothos** [koalaman/shellcheck#2812]
#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:
Consider this function that possibly fails (here with 100% chance) when running with `-e` and where we might want to capture this failure.

```sh
#!/bin/bash
set -euo pipefail

function myf() {
  echo "Entered function"
  false
  echo "SHOULD NOT RUN"
}
```

Normally people would do this: `if ! myf; then echo "function failed"; fi` or `myf || echo "function failed"` (same with `&&` and inversed return code).
However, that's wrong! Both will actually not capture the failure because the function continues to run over `false` and print `SHOULD NOT RUN` and return 0 meaning success. This is because the `if FUNC; then` and `FUNC ||` (or `&&`) constructs somehow overwrite the `-e` trap for their own internal purposes.
The only solution is to run the function outside of these constructs to have `-e` working but on the same time we don't want to terminate the script, so this only runs in a subshell and the main shell temporarily disables `-e`. This here works:
```
set +e
(
  set -e
  myf
)
r=$?
set -e
[ $r -eq 0 ] || echo "Function failed"
# or as if block: if ! [ $r -eq 0 ]; then echo "Function failed"; fi
```

#### Here's what shellcheck currently says:

Nothing

#### Here's what I wanted or expected to see:

A warning or info that suggest the above working invocation.
It's quite some boilerplate code that even can't be moved to a function itself and since bash has no marcos that has to be repeated each time.
But this error is so subtle and strange, it's really something that should be warned about: When `-e` is expected to work, don't call functions in `if` statements or with `||`/`&&`.

**Comment by wileyhy** [comment]
Hi,
  One of the reasons, in my understanding, for the test-commands of an if
construct to have an exemption from the errexit setting is so that the
appropriate choice (then or else {or elif}) can be found. If the choices
were only either "true" or "false and exit," then how would an if construct
be any different from a regular "simple command" also under `set -e`?

Wiley

**Comment by pothos** [comment]
It's already supported but one needs to pass `-o all`. I would say that this could be enabled by default but anyway, closing.
